src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -2593,6 +2607,208 @@ public override IOperation VisitLock(ILockOperation operation, int? captureIdFor
             return null;
         }
 
+        public override IOperation VisitForEachLoop(IForEachLoopOperation operation, int? captureIdForResult)
+        {
+            Debug.Assert(_currentStatement == operation);
+
+            ForEachLoopOperationInfo info = ((BaseForEachLoopStatement)operation).Info;
+
+            bool haveLocals = !operation.Locals.IsEmpty;
+            bool createdRegionForCollection = false;
+
+            if (haveLocals && operation.LoopControlVariable.Kind == OperationKind.VariableDeclarator)
+            {
+                // VB has rather interesting scoping rules for control variable.
+                // It is in scope in collection expression. However, it is considered to be 
+                // "a different" version of that local. Effectively when the code is emitted,
+                // there are two distinct locals, one is used in collection expression and the 
+                // other is used as a loop control variable. This is done to have proper hoisting 
+                // and lifetime in presence of lambdas.
+                // Rather than introducing a separate local symbol, we will simply add another 
+                // lifetime region for that local around the collection expression. 
+
+                var declarator = (IVariableDeclaratorOperation)operation.LoopControlVariable;
+                ILocalSymbol local = declarator.Symbol;
+
+                if (operation.Collection.DescendantsAndSelf().OfType<ILocalReferenceOperation>().Where(l => l.Local.Equals(local)).Any())
```

Comment:
>OfType<ILocalReferenceOperation>().Where(l => l.Local.Equals(local)).Any() [](start = 62, length = 74)

Consider merging some of these. Perhaps: `Any(d => d is ILocalReferenceOperation l && l.Local.Equals(local))`